Add Vault Database Guide (#8747)
quarkusio/quarkus: docs/src/main/asciidoc/vault-datasource.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Using HashiCorp Vault with Databases

include::./attributes.adoc[]
:extension-status: preview
:base-guide: link:vault[Vault guide]
:root-token: s.5VUS8pte13RqekCB2fmMT3u2
:config-file: application.properties

The most common use case when working with Vault is to keep confidential the database connection credentials.
There are several approaches that are supported in Quarkus, with different levels of sophistication

> and security :

Reviewer: ```suggestion
and security:
```

> * Property fetched from the KV Secret Engine using the Vault MP Config Source

Reviewer: ```suggestion
* Property fetched from the KV Secret Engine using the Vault MicroProfile Config Source
```

        return (List<Gift>) em.createQuery("select g from Gift g").getResultList();
    }
}
----

With its `Gift` entity:

[source, java]
----
@Entity
public class Gift {

    private Long id;
    private String name;

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator="giftSeq")
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}
----

Finally, add a new endpoint in `GreetingResource`:

[source, java]
----
@Inject
SantaClausService santaClausService;

@GET
@Path("/gift-count")
@Produces(MediaType.TEXT_PLAIN)
public int geGiftCount() {
    return santaClausService.getGifts().size();
}
----

Start a PostgreSQL database:

[source, shell]
----
docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 --name postgres-quarkus-hibernate -e POSTGRES_USER=sarah -e POSTGRES_PASSWORD=connor -e POSTGRES_DB=mydatabase -p 5432:5432 postgres:10.5
----

Now we are ready to configure Vault and Quarkus to be able to connect to this database from the application.

> == Vault MP Config Source

Reviewer: ```suggestion
== Vault MicroProfile Config Source
```

export VAULT_TOKEN={root-token}
----

Add a `dbpassword` property in the `config` path of the KV secret engine, beside the original `a-private-key`
property:
[source, shell]
----
vault kv put secret/myapps/vault-quickstart/config a-private-key=123456 dbpassword=connor
----

Add the following configuration in Quarkus to use the value of property `dbpassword` as our database password:
[source, properties]
----
# configure your datasource
quarkus.datasource.db-kind = postgresql
quarkus.datasource.username = sarah
quarkus.datasource.password = ${dbpassword}
quarkus.datasource.jdbc.url = jdbc:postgresql://localhost:5432/mydatabase

# drop and create the database at startup (use `update` to only update the schema)
quarkus.hibernate-orm.database.generation=drop-and-create
----

Compile and start the application:
[source, shell]
----
./mvnw package
java -jar target/vault-quickstart-1.0-SNAPSHOT-runner.jar
----

Test it with the `gift-count` endpoint:

[source, shell]
----
curl http://localhost:8080/hello/gift-count
----

You should see:

[source, shell]
----
0
----

This approach is certainly the simplest of all. It has also the big advantage of working with any subsystem

> that requires a secret information in the configuration (i.e. not just _Agroal)_.

Reviewer: ```suggestion
that requires a secret information in the configuration (i.e. not just _Agroal_).
```

The only drawback is that the password will never be fetched again from Vault after the initial property loading.
This means that if the db password was changed while running, the application would have to be restarted after

> vault has been updated with the new password.

Reviewer: ```suggestion
Vault has been updated with the new password.
```

This contrasts with the credentials provider approach, which fetches the password from Vault every time a connection
creation is attempted.

== Credentials Provider

In this approach we introduce a new abstraction called the _Credentials Provider_ that acts as an intermediary

> component between the `agroal` datasource and Vault. The additional configuration required is small, and it

Reviewer: ```suggestion
component between the Agroal datasource and Vault. The additional configuration required is small, and it
```

> is has the major advantage of handling gracefuly database password change while the application is running,

Reviewer: ```suggestion
has the major advantage of handling gracefully database password change while the application is running,
```

quarkus.datasource.jdbc.url = jdbc:postgresql://localhost:5432/mydatabase
----

Recompile, start and test the `gift-count` endpoint. You should see `0` again.

== Dynamic Database Credentials

The two previous approaches work well and are very popular. However they rely on a well known user configured
in the application (i.e. the database user), and the security comes from the confidentiality of the password.
If the password was stolen, we would have to change it in the database and in Vault. Regulary rotating passwords
is actually a very good practice to limit (in time) the impact of getting the password stolen.

A more sophisticated approach consists in letting Vault create and retire database accounts on a regular basis.
This is supported in Vault with the https://www.vaultproject.io/docs/secrets/databases[Database secret engine]. A number
of databases are supported, such as https://www.vaultproject.io/docs/secrets/databases/postgresql[PostgreSQL].

First we need to enable the `database` secret engine, configure the `postgresql-database-plugin` and create
the database role `mydbrole` (replace `10.0.0.3` by the actual host that is running the PostgreSQL container;

> for simplicity we disabled _tls_ between Vault and the PostgreSQL database):

Reviewer: ```suggestion
for simplicity we disabled _TLS_ between Vault and the PostgreSQL database):
```

EOF

vault write database/roles/mydbrole \
    db_name=mydb creation_statements=@vault-postgres-creation.sql \
    default_ttl=1h \
    max_ttl=24h \
    revocation_statements="ALTER ROLE \"{{name}}\" NOLOGIN;" \
    renew_statements="ALTER ROLE \"{{name}}\" VALID UNTIL '{{expiration}}';"
----

[NOTE]
====
For this use case, user `sarah` configured above needs to be a PostgreSQL super user with the capability
to create users.
====

Then we need to give a read capability to the Quarkus application on path `database/creds/mydbrole`.
[source, shell]
----
cat <<EOF | vault policy write vault-quickstart-policy -
path "secret/data/myapps/vault-quickstart/*" {
  capabilities = ["read"]
}
path "database/creds/mydbrole" {
  capabilities = [ "read" ]
}
EOF
----

Now that Vault knows how to create users in PostgreSQL, we juste need to change the `mydatabase` credentials
provider to use a `database-credentials-role`.
[source, properties]
----
quarkus.vault.credentials-provider.mydatabase.database-credentials-role=mydbrole
----

[NOTE]
====
When using `quarkus.hibernate-orm.database.generation=drop-create`, objects get created with the applicative

> user. Since a user will be created everytime the applications starts, database objects will be created

Reviewer: ```suggestion
user. Since a user will be created every time the applications starts, database objects will be created
```
